src/device/dd/disk.c

```diff
@@ -197,3 +201,187 @@ uint32_t PhysToLBA(const struct dd_disk* disk, uint16_t head, uint16_t track, ui
+    uint16_t sectorsize = host_secbyte + 1;
+    uint16_t lba = PhysToLBA(disk, head, track, block);
+
+    if (lba > MAX_LBA)
+    {
+        error = 1;
+        DebugMessage(M64MSG_ERROR, "Invalid LBA (Head:%d - Track:%04x - Block:%d)", head, track, block);
+        return error;
+    }
+
+    const struct dd_sys_data* sys_data = (void*)(disk->istorage->data(disk->storage) + disk->offset_sys);
+    //*bm_zone = LBAToVZone(sys_data, lba);
+    *bm_track_offset = LBAToByte(sys_data, 0, lba) + sector * sectorsize;
+
+    //Handle Errors for wrong System Data
+    if (sector == 0)
+    {
+        uint16_t block = *bm_track_offset / BLOCKSIZE(0);
+        uint16_t block_sys = disk->offset_sys / BLOCKSIZE(0);
+        uint16_t block_id = disk->offset_id / BLOCKSIZE(0);
+
+        if ((block < PROTECT_LBA && block != block_sys)
+         || (block > PROTECT_LBA && block < (DISKID_LBA + 2) && block != block_id)) {
+            error = 1;
+        }
+    }
+
+    if (lba <= MAX_LBA && sector == 0)
+        DebugMessage(M64MSG_VERBOSE, "LBA %d - Offset %08X - Size %04X", lba, *bm_track_offset, sectorsize * SECTORS_PER_BLOCK);
+
+    return error;
+}
+
+static unsigned int seek_track_d64(const struct dd_disk* disk,
+    uint32_t cur_tk, uint32_t cur_sec, uint32_t host_secbyte,
+    unsigned char bm_write, unsigned char bm_block,
+    unsigned int* bm_zone, unsigned int* bm_track_offset)
+{
+    unsigned int error = 0;
+    const struct dd_sys_data* sys_data = (void*)(disk->istorage->data(disk->storage) + disk->offset_sys);
+
+    uint16_t rom_lba_end = big16(sys_data->rom_lba_end);
+    uint16_t ram_lba_start = big16(sys_data->ram_lba_start);
+    uint16_t ram_lba_end = big16(sys_data->ram_lba_end);
+    uint8_t disk_type = sys_data->type & 0x0F;
+
+    uint16_t head = ((cur_tk & 0x1000) / 0x1000);
+    uint16_t track = (cur_tk & 0x0fff);
+    uint16_t block = bm_block;
+    uint16_t sector = cur_sec - bm_write;
+    uint16_t sectorsize = host_secbyte + 1;
+    uint16_t lba = PhysToLBA(disk, head, track, block);
+
+    if (lba < DISKID_LBA)
+    {
+        //System Data
+        *bm_track_offset = disk->offset_sys;
+    }
+    else if ((lba >= DISKID_LBA) && (lba < SYSTEM_LBAS))
+    {
+        //Disk ID
+        *bm_track_offset = disk->offset_id;
+    }
+    else if (lba <= (rom_lba_end + SYSTEM_LBAS))
+    {
+        //ROM Area
+        *bm_track_offset = D64_OFFSET_DATA + LBAToByteA(disk_type, SYSTEM_LBAS, lba - SYSTEM_LBAS) + (sector * sectorsize);
+    }
+    else if (((lba - SYSTEM_LBAS) >= ram_lba_start) && ((lba - SYSTEM_LBAS) <= ram_lba_end))
+    {
+        //RAM Area
+        *bm_track_offset = D64_OFFSET_DATA + LBAToByteA(disk_type, SYSTEM_LBAS, rom_lba_end + 1);
+        *bm_track_offset += LBAToByteA(disk_type, ram_lba_start + SYSTEM_LBAS, lba - SYSTEM_LBAS - ram_lba_start) + (sector * sectorsize);
+    }
+    else
+    {
+        //Invalid
+        error = 1;
+        *bm_track_offset = 0xFFFFFFFF;
```

Comment:
Only this format  sets bm_track_offset = -1 on error, other do not.
Is this normal ?

Reply:
Because D64 format does not contain a full disk's worth of data, so there are potential holes. It's pretty unnecessary that said and an error should be enough. I may have taken this from Project64 as is and didn't think much of it.